### PR TITLE
Delete -fms-extensions from coreclr native build

### PR DESCRIPTION
**Pull request by am11**
For stdcpp conformance, remove the remaining special extensions.
* clean up unnecessary __llvm  special handling.
* Add class, enum, struct prefixes to cases where the variable name is clashing with type name and it changes meaning  (-Werror=changes-meaning)
  * In some trivial cases, I just renamed the variables..
* On x86, replace `__asm {}` syntax with `__asm ("")`
* Two small illumos related build fixes under `src/native` which helped validating the rest of the changes on the platform.

### CODE REVIEW

File: src/coreclr/utilcode/md5.cpp [hunk 1]
```diff
@@ -141,11 +142,10 @@ void MD5::GetHashValue(MD5HASHDATA* phash)
     //
     // but our compiler has an intrinsic!
 
-    #if (defined(HOST_X86) || defined(HOST_ARM) || !defined(__clang__)) && defined(TARGET_UNIX)
-    #define ROL(x, n)        (((x) << (n)) | ((x) >> (32-(n))))
-    #define ROTATE_LEFT(x,n) (x) = ROL(x,n)
+    #if __has_builtin(__builtin_rotateleft)
```
Comment (am11) [inline src/coreclr/utilcode/md5.cpp:145]:
```suggestion
    #if defined(TARGET_WINDOWS)
    #define ROTATE_LEFT(x, n) (x) = _lrotl(x, n)
    #elif __has_builtin(__builtin_rotateleft)
```

File: src/coreclr/utilcode/md5.cpp [hunk 2]
```diff
@@ -141,11 +142,12 @@ void MD5::GetHashValue(MD5HASHDATA* phash)
     //
     // but our compiler has an intrinsic!
 
-    #if (defined(HOST_X86) || defined(HOST_ARM) || !defined(__clang__)) && defined(TARGET_UNIX)
-    #define ROL(x, n)        (((x) << (n)) | ((x) >> (32-(n))))
-    #define ROTATE_LEFT(x,n) (x) = ROL(x,n)
+    #if defined(TARGET_WINDOWS)
+    #define ROTATE_LEFT(x, n) (x) = _lrotl(x, n)
```
Comment (am11) [inline src/coreclr/utilcode/md5.cpp:146]:
```suggestion
        #define ROTATE_LEFT(x, n) (x) = _lrotl(x, n)
```

File: src/coreclr/utilcode/stacktrace.cpp [hunk 3]
```diff
@@ -955,7 +955,13 @@ void MagicDeinit(void)
 *       Exactly the contents of RtlCaptureContext for Win7 - Win2K doesn't
 *       support this, so we need it for CoreCLR 4, if we require Win2K support
 ****************************************************************************/
-extern "C" __declspec(naked) void __stdcall
+extern "C"
+#ifdef TARGET_WINDOWS
+__declspec(naked)
+#else
+__attribute__((naked))
+#endif
```
Comment (huoyaoyuan) [inline src/coreclr/utilcode/stacktrace.cpp:963]: Do we need a macro for this? Like NOINLINE
Or for naked pure ASM functions, does it make sense to move the implementation to asm file?

File: src/native/libs/System.Native/pal_networking.c [hunk 4]
```diff
@@ -2005,9 +2005,11 @@ static bool TryGetPlatformSocketOption(int32_t socketOptionLevel, int32_t socket
                     *optName = TCP_KEEPINTVL;
                     return true;
 
+#ifdef TCP_FASTOPEN_CONNECT
```
Comment (huoyaoyuan) [inline src/native/libs/System.Native/pal_networking.c:2008]: Why is this included?

File: src/coreclr/utilcode/stacktrace.cpp (hunk 3 above)
Comment (am11) [reply src/coreclr/utilcode/stacktrace.cpp:963]: The only common denominator in the few usages of naked attribute is minipal, so I just inlined it. This is x86 specific hack. Most likely it can be replaced by C/C++ counterparts instead of keeping it inlined.

File: src/native/libs/System.Native/pal_networking.c (hunk 4 above)
Comment (am11) [reply src/native/libs/System.Native/pal_networking.c:2008]: Yes, it was last week's change https://github.com/am11/runtime/commit/e377f169c0d790172baf762f93dc011f727777e5, the other usage was guarding with this macro, this was an oversight (CI doesn't have anything which was testing these #else cases).

File: src/coreclr/vm/eventtrace.cpp [hunk 5]
```diff
@@ -227,10 +227,18 @@ extern "C"
     CallStackFrame* GetEbp()
     {
         CallStackFrame *frame=NULL;
+#ifdef TARGET_WINDOWS
         __asm
         {
             mov frame, ebp
         }
+#else
+        __asm
```
Comment (janvorli) [inline src/coreclr/vm/eventtrace.cpp:236]: `__builtin_frame_address(0)` intrinsic on Unix should do the same thing

File: src/coreclr/vm/spinlock.h [hunk 6]
```diff
@@ -112,11 +112,11 @@ class DangerousNonHostedSpinLock
         return (BOOL)m_value;
     }
 
-    typedef Holder<DangerousNonHostedSpinLock *, DangerousNonHostedSpinLock::AcquireLock, DangerousNonHostedSpinLock::ReleaseLock> Holder;
+    typedef Holder<DangerousNonHostedSpinLock *, DangerousNonHostedSpinLock::AcquireLock, DangerousNonHostedSpinLock::ReleaseLock> Holder1;
```
Comment (janvorli) [inline src/coreclr/vm/spinlock.h:115]: A nit - how about renaming it this way instead:
```
Holder->LockHolder
TryHolder->LockTryHolder
```

File: src/coreclr/vm/eventtrace.cpp [hunk 7]
```diff
@@ -227,10 +227,18 @@ extern "C"
...
+        (
+            "mov %%ebp, %0\n\t"
+            : "=r" (frame)
+        );
```
Comment (am11) [inline src/coreclr/vm/eventtrace.cpp:240]:
```suggestion
        frame = (CallStackFrame*) __builtin_frame_address(0);
```

File: src/coreclr/nativeaot/Runtime/eventpipe/ds-rt-aot.cpp [hunk 8]
```diff
@@ -56,7 +56,7 @@ aot_ipc_get_process_id_disambiguation_key(
 	// since the start of the Unix epoch).
 	struct kinfo_proc info = {};
 	size_t size = sizeof (info);
-	int mib [4] = { CTL_KERN, KERN_PROC, KERN_PROC_PID, process_id };
+	int mib [4] = { CTL_KERN, KERN_PROC, KERN_PROC_PID, (int32_t)process_id };
```
[the remark this answers is not in the document]
Comment (am11) [reply src/coreclr/nativeaot/Runtime/eventpipe/ds-rt-aot.cpp:59]:
```suggestion
	int mib [4] = { CTL_KERN, KERN_PROC, KERN_PROC_PID, (int)process_id };
```

File: src/coreclr/pal/src/thread/process.cpp [hunk 9]
```diff
@@ -1629,7 +1629,7 @@ GetProcessIdDisambiguationKey(DWORD processId, UINT64 *disambiguationKey)
     // since the start of the Unix epoch).
     struct kinfo_proc info = {};
     size_t size = sizeof(info);
-    int mib[4] = { CTL_KERN, KERN_PROC, KERN_PROC_PID, processId };
+    int mib[4] = { CTL_KERN, KERN_PROC, KERN_PROC_PID, (int32_t)processId };
```
[the remark this answers is not in the document]
Comment (am11) [reply src/coreclr/pal/src/thread/process.cpp:1632]:
```suggestion
    int mib[4] = { CTL_KERN, KERN_PROC, KERN_PROC_PID, (int)processId };
```

File: src/coreclr/utilcode/md5.cpp [hunk 10]
```diff
@@ -141,11 +142,12 @@ void MD5::GetHashValue(MD5HASHDATA* phash)
...
-    #if (defined(HOST_X86) || defined(HOST_ARM) || !defined(__clang__)) && defined(TARGET_UNIX)
-    #define ROL(x, n)        (((x) << (n)) | ((x) >> (32-(n))))
-    #define ROTATE_LEFT(x,n) (x) = ROL(x,n)
+    #if defined(TARGET_WINDOWS)
+        #define ROTATE_LEFT(x, n) (x) = _lrotl(x, n)
+    #elif __has_builtin(__builtin_rotateleft)
+        #define ROTATE_LEFT(x, n) ((x) = __builtin_rotateleft(x, n))
     #else
-    #define ROTATE_LEFT(x,n) (x) = _lrotl(x,n)
+        #define ROTATE_LEFT(x, n) ((x) = ((x << (n)) | (x >> ((sizeof(x) * 8 - 1) & -(n)))))
```
Comment (janvorli) [inline src/coreclr/utilcode/md5.cpp:150]: I wonder what is the benefit of using `((sizeof(x) * 8 - 1) & -(n)))` as a shift amount instead of an easier to read `(sizeof(x) * 8 - (n))`.

Comment (am11) [reply src/coreclr/utilcode/md5.cpp:150]:
```suggestion
        #define ROTATE_LEFT(x, n) ((x) = ((x << (n)) | (x >> (sizeof(x) * 8 - (n)))))
```

File: src/coreclr/tools/superpmi/superpmi-shared/standardpch.h [hunk 11]
```diff
@@ -110,13 +110,6 @@
 #define DEFAULT_REAL_JIT_NAME_A MAKEDLLNAME_A("clrjit2")
 #define DEFAULT_REAL_JIT_NAME_W MAKEDLLNAME_W("clrjit2")
 
-#if !defined(_MSC_VER) && !defined(__llvm__)
```
[the remark this answers is not in the document]
Comment (janvorli) [reply src/coreclr/tools/superpmi/superpmi-shared/standardpch.h:113]: I was rather concerned about the change on Windows due to the fact that you've replaced `__debugbreak` at multiple places by `DebugBreak`.

[the remark this answers is not in the document]
Comment (am11) [reply src/coreclr/tools/superpmi/superpmi-shared/standardpch.h:113]: Yup, I will change it shortly (that's why it is unresolved 🙂).

File: src/coreclr/vm/common.h [hunk 12]
```diff
@@ -204,11 +204,7 @@ EXTERN_C AppDomain* STDCALL GetAppDomain();
 
 inline void RetailBreak()
 {
-#ifdef TARGET_X86
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/vm/common.h:207]: Historically, inline `int 3` gave a better experience under the debugger - the debugger stopped in your code, not inside a Windows `DebugBreak` API.

File: src/coreclr/vm/common.h [hunk 13]
```diff
@@ -204,11 +204,7 @@ EXTERN_C AppDomain* STDCALL GetAppDomain();
 
 inline void RetailBreak()
```
Comment (jkotas) [inline src/coreclr/vm/common.h:205]: I do not see this method being used anywhere. Delete it?

File: src/coreclr/tools/superpmi/superpmi-shared/standardpch.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (am11) [reply src/coreclr/tools/superpmi/superpmi-shared/standardpch.h:113]: Defined and use a macro:

```c
#ifdef _MSC_VER
#define DEBUG_BREAK __debugbreak()
#else
#define DEBUG_BREAK DebugBreak()
#endif
```

File: src/coreclr/minipal/minipal.h [hunk 14]
```diff
@@ -3,6 +3,12 @@
 //
 #include <stddef.h>
 
+#ifdef _MSC_VER
+#define DEBUG_BREAK __debugbreak()
+#else
+#define DEBUG_BREAK DebugBreak()
```
Comment (jkotas) [inline src/coreclr/minipal/minipal.h:9]: `DebugBreak()` is defined in CoreCLR Win32 PAL. minipal should not have a dependency on CoreCLR Win32 PAL. This macro should be in CoreCLR Win32 PAL. Or even - since this seems to be only used in superpmi test tool, define it for superpmi only.

This macro is not very useful. Release build should not contain debug breaks. And asserts provide equivalent for debug builds.

Comment (am11) [reply src/coreclr/minipal/minipal.h:9]: Ah, I actually have it defined in `src/coreclr/tools/superpmi/superpmi-shared/standardpch.h`. First put it there to see if it compiles then forgot to revert this file.

File: src/coreclr/minipal/minipal.h [hunk 15]
```diff
@@ -3,6 +3,12 @@
...
+#endif
+
```
Comment (am11) [inline src/coreclr/minipal/minipal.h:11]:
```suggestion
```

File: src/native/libs/System.Native/pal_networking.c [hunk 16]
```diff
@@ -2015,9 +2015,11 @@ static bool TryGetPlatformSocketOption(int32_t socketOptionLevel, int32_t socket
...
```
Comment (am11) [inline src/native/libs/System.Native/pal_networking.c:2018]:
```suggestion
#ifdef TCP_FASTOPEN
```

**Comment by am11** [comment]
OSX failure is infra issue (tests are passing but helix job is unable to exit cleanly):
```
/private/tmp/helix/working/A8E70893/w/A68408BE/e /private/tmp/helix/working/A8E70893/w/A68408BE/e
  Discovering: System.Text.RegularExpressions.Unit.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Text.RegularExpressions.Unit.Tests (found 25 test cases)
  Starting:    System.Text.RegularExpressions.Unit.Tests (parallel test collections = on [12 threads], stop on fail = off)
  Finished:    System.Text.RegularExpressions.Unit.Tests
=== TEST EXECUTION SUMMARY ===
   System.Text.RegularExpressions.Unit.Tests  Total: 830, Errors: 0, Failed: 0, Skipped: 0, Time: 0.352s
/private/tmp/helix/working/A8E70893/w/A68408BE/e
----- end Wed Jun 12 17:28:07 EDT 2024 ----- exit code 0 ----------------------------------------------------------
exit code 0 means Exited Successfully
Unable to obtain kernel buffer: Operation not permitted
usage: sudo dmesg
----- start =============== XUnitLogChecker Output =====================================================
No dumps found in /cores.
```
rest of failures are known issues (per Build Analysis).

@jkotas, PTAL.

**Review by jkotas — APPROVED** [review]
Thank you!
